File: 079 Word Search/WordSearch.java
class Solution {
    public boolean exist(char[][] board, String word) {  
        char[] w = word.toCharArray();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (exist(board, i, j, w, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
    
    private boolean exist(char[][] board, int i, int j, char[] word, int a) {
        if (a == word.length) {
            return true;
        }
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) {
            return false;
        }
        if (board[i][j] != word[a]) {
            return false;
        }
        
        board[i][j] ^= 256; // bit mask
        boolean existed = exist(board, i + 1, j, word, a + 1) || 
            exist(board, i - 1, j, word, a + 1) ||
            exist(board, i, j + 1, word, a + 1) ||
            exist(board, i, j - 1, word, a + 1);
        board[i][j] ^= 256; // recover
        return existed;
    }
}

File: 079 Word Search/word_search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(board, i, j, word, 0):
                    return True
        return False
                
    def dfs(self, board, i, j, word, a):
        if a == len(word):
            return True
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
            return False
        if board[i][j] != word[a]:
            return False
        
        temp = board[i][j]
        board[i][j] = False
        existed = self.dfs(board, i + 1, j, word, a + 1) or self.dfs(board, i - 1, j, word, a + 1) or \
        self.dfs(board, i, j + 1, word, a + 1) or self.dfs(board, i, j - 1, word, a + 1)
        board[i][j] = temp
        return existed